test/TestAppUtils/VisualTreeDumper.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Media;
+
+namespace MUXControls.TestAppUtils
+{
+    public class VisualTreeDumper
+    {
+        public interface IFilter
+        {
+            bool ShouldLogElement(string elementName);
+            bool ShouldLogProperty(string propertyName);
+            bool IsKnownProperty(string propertyName, string value);
+        }
+
+        public interface IPropertyValueTranslator
+        {
+            string PropertyValueToString(string propertyName, Object value);
+        }
+
+        interface IVisitor
+        {
+            bool ShouldVisitNode(DependencyObject node);
+            void BeginVisitNode(DependencyObject node);
+            void EndVisitNode(DependencyObject node);
+
+            bool ShouldVisitPropertiesForNode(DependencyObject node);
+            bool ShouldVisitProperty(PropertyInfo propertyInfo);
+            void VisitProperty(String propertyName, Object value);
+        }
+
+        class Visitor : IVisitor
+        {
+            private StringBuilder _sb;
+            private int _indent;
+            private IFilter _filter;
+            private IPropertyValueTranslator _translator;
+            public Visitor(IFilter filter, IPropertyValueTranslator translator)
+            {
+                _sb = new StringBuilder();
+                _sb.AppendLine("<?xml version=\"1.0\" encoding=\"utf-8\" ?>");
+                _indent = 0;
+                _filter = filter;
+                _translator = translator;
+            }
+            public void EndVisitNode(DependencyObject obj)
+            {
+                _indent--;
+                AddPadding(_indent);
+                _sb.AppendLine("</Element>");
+            }
+
+            public void BeginVisitNode(DependencyObject obj)
+            {
+                AddPadding(_indent);
+                _sb.AppendLine(String.Format("<Element Type=\"{0}\">", obj.GetType().FullName));
+                _indent++;
+            }
+
+            public override String ToString()
+            {
+                return _sb.ToString();
+            }
+
+            public bool ShouldVisitNode(DependencyObject node)
+            {
+                return node != null && _filter.ShouldLogElement(node.GetType().FullName);
+            }
+
+            public bool ShouldVisitPropertiesForNode(DependencyObject node)
+            {
+                return (node as UIElement) != null && _filter.ShouldLogElement(node.GetType().FullName);
```

Comment:
Are we walking over non UIElements?

Reply:
It's from https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.visualtreehelper.getchild, they are dependencyobject.
Also I see non UIelement before but I can't remember the detail